src/qibolab/instruments/zhinst/components/channel.py

```diff
@@ -0,0 +1,19 @@
+from dataclasses import dataclass
+
+from qibolab.components import Channel
+
+__all__ = [
+    "ZiChannel",
+]
+
+
+@dataclass(frozen=True)
+class ZiChannel:
+    """Channel for Zurich Instruments (ZI) devices."""
+
+    logical_channel: Channel
+    """Corresponding logical channel."""
+    device: str
+    """Name of the device."""
+    path: str
+    """Path of the device node."""
```

Comment:
I this is exactly what was done by the plethora of Port classes in the past, and I think it didn't work out that well. The main idea in this proposal is that each instrument defines their own channels however they want, without any restriction. Instrument channels are not supposed to be exposed to users - only respective driver knows what it is, so this should be fine.

Reply:
Well, the plethora of `Port` classes is going to be replaced by a swarm of Channels.

What I'm proposing above is to avoid any driver specific object (so it can not be the same).
If the channels are not user facing, they are not going to exist in the public API, so the information is being conveyed in some other way. And if the user has enough room to pass this information without them, we should be able to deal with everything without them.

Essentially, we already have a group of driver specific objects, the `Instrument`s. So, we should be able to pass all the information that is specific to them in their configurations, and avoid any further specialized object.

Reply:
I think having a specialized channel for each instrument is still useful. For example for QM https://github.com/qiboteam/qibolab/blob/443dac8c8fa5dc5b97867fcc856df1943c78b37c/src/qibolab/instruments/qm/components/channel.py#L11, I introduced a channel that is a bit different than the ZI counterpart, since we are using device name (`str`) and port (`int`) as identifiers, in contrast to the ZI path.

I think the price of separate channels is that the users writing platforms (ie. qibolab_platforms_qrc) need to be familiar with the channel API for the instruments they are using, which is a fair assumption. We could probably propose a unified interface that is the same for all instruments and do the translations internally in qibolab, however this would increase the code complexity and we would not get that much in return.

Reply:
> What I'm proposing above is to avoid any driver specific object (so it can not be the same). If the channels are not user facing, they are not going to exist in the public API, so the information is being conveyed in some other way.

I believe the instrument specific channels are user facing, as they are needed to define platforms, for example:
https://github.com/qiboteam/qibolab/blob/982c843f41630f26a926af7ae2b38e720eb5f6dd/tests/dummy_qrc/zurich/platform.py#L76-L80
Therefore it is important to stabilize this as much as possible, to minimize future breaking changes.

Reply:
> I believe the instrument specific channels are user facing

That was my understanding, but not what @hay-k wrote above:

> Instrument channels are not supposed to be exposed to users

That's why I was arguing.

> We could probably propose a unified interface that is the same for all instruments and do the translations internally in qibolab, however this would increase the code complexity and we would not get that much in return.

About this I do not fully agree: instead of the device-specific class, we would have a device-specific conversion, of objects that should somehow be similar.
The tradeoff is set by how similar are the objects, and how much information should be encoded in there. Maybe we could just discuss the alternatives (even live), also because this is the current difference I see:
```py
@dataclass(frozen=True)
class QmChannel:
    """Channel for Zurich Instruments (ZI) devices."""

    logical_channel: Channel
    """Corresponding logical channel."""
    device: str
    """Name of the device."""
    port: int
    """Number of port."""


@dataclass(frozen=True)
class ZiChannel:
    """Channel for Zurich Instruments (ZI) devices."""

    logical_channel: Channel
    """Corresponding logical channel."""
    device: str
    """Name of the device."""
    path: str
    """Path of the device node."""
```
that to me seems the exact same code, just that in one case the `port` is an integer, and in the other there is a more generic `path` (which you can easily consider as a superset of integers, through the trivial `int()` encoding).

Reply:
(btw, they are so close to each other that the docstring says ZI for both :P)

Reply:
> Well, the plethora of Port classes is going to be replaced by a swarm of Channels.

With ports it was different. There was an attempt to have a unified handling of them, hence there was an attempt to introduce a parent class Port, but all child classes had nothing to do with the parent. With device specific channels there is no attempt to generalize anything.

> I believe the instrument specific channels are user facing, as they are needed to define platforms

User: person who runs calibration experiments, or quantum algorithms. What user does is repetitive and never ending activity.

Administrator / User level God: person(s) who sets up the platform so that users can use it. What administrator does is done once and may need occasional adjustment over time.

In some cases the two are the same person, in which case it is their responsibility to know knowledge about both worlds. Users calibrating or running quantum circuits do not need to know platform details, hence those are not exposed to them. Administrator needs to know every single detail, hence instrument specific channels are for them.

Reply:
Well, I see your point in splitting the two kind of users. But my definition was easier (and I still believe the most common one), i.e. whoever is invoking Qibolab functions outside Qibolab itself.

In any case, it is not much relevant to discuss the *user* word definition.
Instead, the substance is that Qibolab development is a separate process from Qibolab usage, and the distinction is about what is part of the public API and what not.
Someone writing a platform has to know the details about the instrument. But it doesn't have to know anything about Qibolab's code, other than what is explained in the official docs.
Having a single channel means that it's easier to move from one instrument to another, and it would avoid device-specific docs for Qibolab. If possible, it's simpler in many respects.

What I'm proposing here is just to replace the two classes above with:
```py
@dataclass(frozen=True)
class Channel:
    logical_channel: Channel
    """Corresponding logical channel."""
    device: str
    """Name of the device."""
    path: str
    """Internal address."""
```
and to interpret within QM `path` as the port number with `int(ch.path)`.

I don't see any further information that should belong to one driver and not the other. And even in the case in which there would be something like that, related to a feature truly available on one device only (not just a different encoding), we could keep spare attributes, until they are a minority (for the time being is this minority amounts to zero).

Reply:
> Well, I see your point in splitting the two kind of users. But my definition was easier (and I still believe the most common one), i.e. whoever is invoking Qibolab functions outside Qibolab itself.

I also have this definition in mind. Ideally, if Qibolab is simple enough, documented and more importantly stable, we shouldn't need administrator level users to write the platforms. People in the lab connecting them (= standard users) should be able to do it themselves.

> What I'm proposing here is just to replace the two classes above with:
> 
> ```python
> @dataclass(frozen=True)
> class Channel:
>     logical_channel: Channel
>     """Corresponding logical channel."""
>     device: str
>     """Name of the device."""
>     path: str
>     """Internal address."""
> ```
> 
> and to interpret within QM `path` as the port number with `int(ch.path)`.

This sounds like a reasonable suggestion. It would certainly work for ZI and QM. I am not sure about other instruments, but I would expect it would, since `str` can be easily used to represent different things. I believe we could give a try having a single `Channel`, it would also simplify writing platforms hopefully making it more accessible to (non-administrator) users. We can still keep the `Config`s instrument specific.